smolyay/surrogate.py

```diff
@@ -0,0 +1,272 @@
+        self._reset_grid()
+
+    @property
+    def grid_generator(self):
+        """:class:`IndexGridGenerator`: Generator used for making the grids."""
+        return self._grid_generator
+
+    @grid_generator.setter
+    def grid_generator(self, value):
+        if not isinstance(value, IndexGridGenerator):
+            raise TypeError('Grids must be IndexGridGenerator')
+        self._grid_generator = value
+        self._reset_grid()
+
+    @property
+    def grid(self):
+        """:class:`IndexGrid` data structure."""
+        if self._grid is None:
+            self._coefficients = None
+            self._grid = self.grid_generator(self.domain.shape[0])
+        return self._grid
+
+    def _reset_grid(self):
+        """Reset the points and basis functions if the generator is changed."""
+        self._grid = None
+
+    def _make_basis_matrix(self):
+        """Generate basis matrix.
+
+        Parameters
+        ----------
+        points: list
+            Grid points.
+        basis_functions: list
+            Basis function of grid points.
+
+        Returns
+        -------
+        numpy.ndarray
+            Basis matrix.
+        """
+        points, basis_functions = self.grid.points, self.grid.basis_functions
+        basis_matrix = numpy.zeros((len(points),
+                                    len(points)))
+        for i, point in enumerate(points):
+            for j, basis in enumerate(basis_functions):
+                basis_matrix[i, j] = numpy.prod(
+                    [f(x) for x, f in zip(point, basis)])
+        return basis_matrix
+
+    def train(self, real_function, linear_solver='lu'):
+        """Fit surrogate's components (basis functions) to the function.
+
+        Parameters
+        ----------
+        real_function: callabe
+            Function to be approximated.
+        linear_slover: string
+            Method of solving linear equations:
+                'lu': lower upper decomposition.
+                'inv': solve for coefficients through x = A^{-1}B.
+                'lstsq': least-square solution.
+
+        Returns
+        -------
+        list
+            Coefficients of the surrogate.
+        """
+        # transform grid points
+        transformed_grids = numpy.zeros((
+            numpy.array(self.grid.points).shape[0], self._domain.shape[0]))
+        for dimension in range(self._domain.shape[0]):
+            transformed_grids[:, dimension] = (
+                numpy.polynomial.polyutils.mapdomain(
+                    numpy.array(self.grid.points)[:, dimension],
+                    (-1, 1), self.domain[dimension]))
+
+        # evaluate real function at grid points
+        real_function_at_grids = numpy.zeros((numpy.array(
+            self.grid.points).shape[0]))
+        for point_num in range(numpy.array(self.grid.points).shape[0]):
+            real_function_at_grids[point_num] = (
+                real_function(*transformed_grids[point_num]))
```

Comment:
You can do this with list comprehension:
```suggestion
        data = [function(point) for point in self.points]
```
In writing this, I am suggesting that you expose a `points` property of the surrogate that gives the points where sampling should be done:
```
@property
def points(self):
    if self._points is None:
        self._points = self._mapdomain(self.grid.points, [[-1, 1]]*self.dimension, self.domain)
    return self._points
```
Make sure to add `self._points = None` to the constructor and the `_reset_grid()` method.